### PR TITLE
Improvements and fixes to BilinPL

### PR DESCRIPTION
**Pull request by mirkojuergens**
This commit will:
* Fix bugs in the bilinPL implementation related to sum-indices
* Refactor the implementation to be compliant with the paper (variable names etc)
* Add gradient descent as an alternative optimizer
* Add a Black-Box gradient implementation such that we can Black-Box optimize any function (or, use it as a backup if our Gradient returns NaN)
* Add javadoc to the algorithms
* Add some convience methods such as the calculation of the likelihood, streaming API to Vectors and direct Typecasts in DyadRankingDataset
* Change the AdvancedDyadRankerTest to a model that BilinPL can actually learn

-------
Please do not merge until all other feature branches are merged as I unfortunately had to move some packages.

### CODE REVIEW

File: JAICore/jaicore-ml/src/jaicore/ml/core/optimizing/graddesc/BlackBoxGradient.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * where x is the provided point and x' is a point that slightly differs
+ * (specified by the parameter <code>precision</code>. (Obviously it holds that
+ * in lim_{precision -> 0} this yields the exact gradient.)
+ * 
+ * If x is a vector (a_o, ..., a_n), then, instead we calculate each partial
+ * derivative i by:
+ * 
+ * (f(a_o, ... a_i +h, ... , a_n) - f((a_o, ..., a_n)))/h
+ * 
+ * Obviously, this is an highly in efficient approach for estimating the
```
Comment (helegraf) [inline JAICore/jaicore-ml/src/jaicore/ml/core/optimizing/graddesc/BlackBoxGradient.java]:
```suggestion
  * Obviously, this is a highly inefficient approach for estimating the
```

File: JAICore/jaicore-ml/src/jaicore/ml/core/optimizing/graddesc/GradientDescentOptimizer.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		return gradients.stream().allMatch(grad -> Math.abs(grad) < gradientThreshold || !Double.isFinite(grad));
+	}
+
+	private void updatePredictions(Vector initialGuess, Vector gradients) {
+		for (int i = 0; i < initialGuess.length(); i++) {
+			double weight = initialGuess.getValue(i);
+			double gradient = gradients.getValue(i);
+			// don't further optimize if we meet the threshold
+			if (Math.abs(gradient) < gradientThreshold)
+				continue;
```
Comment (helegraf) [inline JAICore/jaicore-ml/src/jaicore/ml/core/optimizing/graddesc/GradientDescentOptimizer.java:78]:
```suggestion
 			{
 				continue;
 			}
```

File: JAICore/jaicore-math/src/de/upb/isys/linearalgebra/Vector.java [hunk 3]
```diff
@@ -1,413 +1,425 @@
...
+	 * 
+	 * @return the stream of the values.
+	 */
+	default DoubleStream stream() {
+		return Arrays.stream(this.asArray());
+	}
+
+
+	
+	public static Vector merge(Vector v1, Vector v2) {
```
Comment (helegraf) [inline JAICore/jaicore-math/src/de/upb/isys/linearalgebra/Vector.java:414]: Add javadoc

File: JAICore/jaicore-ml/src/jaicore/ml/core/optimizing/graddesc/BlackBoxGradient.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ * @author Mirko Jürgens
+ *
+ */
+public class BlackBoxGradient implements IGradientFunction {
+
+	private final double precision;
+
+	private final IGradientDescendableFunction function;
+
+	public BlackBoxGradient(IGradientDescendableFunction underlyingFunction, double precision) {
```
Comment (helegraf) [inline JAICore/jaicore-ml/src/jaicore/ml/core/optimizing/graddesc/BlackBoxGradient.java:40]: Add javadoc

File: JAICore/jaicore-ml/src/jaicore/ml/core/optimizing/graddesc/GradientDescentOptimizer.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+	private final double learningRate;
+
+	private final double gradientThreshold;
+
+	private final int maxIterations;
+
+	private static final Logger log = LoggerFactory.getLogger(GradientDescentOptimizer.class);
+
+	public GradientDescentOptimizer(GradientDescentOptimizerConfig config) {
```
Comment (helegraf) [inline JAICore/jaicore-ml/src/jaicore/ml/core/optimizing/graddesc/GradientDescentOptimizer.java:42]: Add javadoc

File: JAICore/jaicore-ml/src/jaicore/ml/core/optimizing/graddesc/GradientDescentOptimizer.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+		this.learningRate = config.learningRate();
+		this.gradientThreshold = config.gradientThreshold();
+		this.maxIterations = config.maxIterations();
+	}
+
+	public GradientDescentOptimizer() {
```
Comment (helegraf) [inline JAICore/jaicore-ml/src/jaicore/ml/core/optimizing/graddesc/GradientDescentOptimizer.java:48]: Add javadoc

File: JAICore/jaicore-ml/src/jaicore/ml/core/optimizing/graddesc/GradientDescentOptimizer.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import de.upb.isys.linearalgebra.Vector;
+import jaicore.ml.core.optimizing.IGradientBasedOptimizer;
+import jaicore.ml.core.optimizing.IGradientDescendableFunction;
+import jaicore.ml.core.optimizing.IGradientFunction;
+
+/**
+ * An optimizer based on the gradient descent method [1].
```
Comment (helegraf) [inline JAICore/jaicore-ml/src/jaicore/ml/core/optimizing/graddesc/GradientDescentOptimizer.java]: What does [1] reference?

File: JAICore/jaicore-ml/src/jaicore/ml/core/optimizing/graddesc/GradientDescentOptimizerConfig.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
+package jaicore.ml.core.optimizing.graddesc;
+
+import org.aeonbits.owner.Config.Sources;
+import org.aeonbits.owner.Mutable;
+
+@Sources({ "file:conf/opt/graddesc.properties" })
+public interface GradientDescentOptimizerConfig extends Mutable{
```
Comment (helegraf) [inline JAICore/jaicore-ml/src/jaicore/ml/core/optimizing/graddesc/GradientDescentOptimizerConfig.java:7]: Add javadoc

File: JAICore/jaicore-ml/test/jaicore/ml/dyadranking/general/DyadRankingInstanceSupplier.java [hunk 9]
```diff
@@ -44,17 +56,35 @@ public int compare(Dyad d1, Dyad d2) {
...
+
+	private static final double bilinFunc(Vector scoreVec1, Vector scoreVec2) {
+		// x1 * y1 + x2*y2 + x1*y2 + x2*y1
+		double score = scoreVec1.getValue(0) * scoreVec2.getValue(0) + scoreVec1.getValue(1) * scoreVec2.getValue(1)
+				+ scoreVec1.getValue(0) * scoreVec2.getValue(1) + scoreVec1.getValue(1) * scoreVec2.getValue(0);
+		return Math.exp(score);
+	}
+
+	/**
+	 *
```
Comment (helegraf) [inline JAICore/jaicore-ml/test/jaicore/ml/dyadranking/general/DyadRankingInstanceSupplier.java:83]: Add what the method does to the javadoc

File: JAICore/jaicore-ml/src/jaicore/ml/dyadranking/algorithm/featuretransform/FeatureTransformPLDyadRanker.java [hunk 10]
```diff
@@ -121,14 +131,51 @@ public void train(IDataset dataset) throws TrainingException {
...
-		int alternativeLength = ((IDyadRankingInstance) dRDataset.get(0)).getDyadAtPosition(0).getAlternative()
-				.length();
-		int instanceLength = ((IDyadRankingInstance) dRDataset.get(0)).getDyadAtPosition(0).getInstance().length();
+		int alternativeLength = dRDataset.get(0).getDyadAtPosition(0).getAlternative().length();
+		int instanceLength = dRDataset.get(0).getDyadAtPosition(0).getInstance().length();
 		Vector initialGuess = new DenseDoubleVector(
-				featureTransform.getTransformedVectorLength(alternativeLength, instanceLength));
-		initialGuess.fillRandomly();
+				featureTransform.getTransformedVectorLength(alternativeLength, instanceLength), -0.3);
+		// initialGuess.fillRandomly();
```
Comment (helegraf) [inline JAICore/jaicore-ml/src/jaicore/ml/dyadranking/algorithm/featuretransform/FeatureTransformPLDyadRanker.java:138]: Why is the Vector not filled randomly anymore? What is the random guess?

File: JAICore/jaicore-ml/src/jaicore/ml/dyadranking/algorithm/IPLNetDyadRankerConfiguration.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+import java.io.File;
+import java.util.List;
+
+import org.aeonbits.owner.Config.Sources;
+
+import jaicore.ml.core.predictivemodel.IPredictiveModelConfiguration;
+
+@Sources({ "file:conf/plNet/plnet.properties" })
+public interface IPLNetDyadRankerConfiguration extends IPredictiveModelConfiguration {
```
Comment (helegraf) [inline JAICore/jaicore-ml/src/jaicore/ml/dyadranking/algorithm/IPLNetDyadRankerConfiguration.java]: Add javadoc

**Review by helegraf — DISMISSED** [review]
Looking good, some minor java doc changes

File: JAICore/jaicore-math/src/de/upb/isys/linearalgebra/Vector.java (hunk 3 above)
Comment (mirkojuergens) [reply JAICore/jaicore-math/src/de/upb/isys/linearalgebra/Vector.java:414]: this wasn't done by me :D

File: JAICore/jaicore-ml/src/jaicore/ml/dyadranking/algorithm/featuretransform/FeatureTransformPLDyadRanker.java (hunk 10 above)
Comment (mirkojuergens) [reply JAICore/jaicore-ml/src/jaicore/ml/dyadranking/algorithm/featuretransform/FeatureTransformPLDyadRanker.java:138]: if we set the inital weight too large we will have a lot of overflows.
